raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2248,24 +2250,23 @@ public void poll() {
     }
 
     @Override
-    public Long scheduleAppend(int epoch, List<T> records) {
+    public long scheduleAppend(int epoch, List<T> records) {
         return append(epoch, records, false);
     }
 
     @Override
-    public Long scheduleAtomicAppend(int epoch, List<T> records) {
+    public long scheduleAtomicAppend(int epoch, List<T> records) {
         return append(epoch, records, true);
     }
 
-    private Long append(int epoch, List<T> records, boolean isAtomic) {
-        Optional<LeaderState<T>> leaderStateOpt = quorum.maybeLeaderState();
-        if (!leaderStateOpt.isPresent()) {
-            return Long.MAX_VALUE;
-        }
+    private long append(int epoch, List<T> records, boolean isAtomic) {
+        LeaderState<T> leaderState = quorum.<T>maybeLeaderState().orElseThrow(
+            () -> new NotLeaderException("Append failed because the replication is not the current leader")
```

Comment:
I think it would be nice if we could consolidate `NotLeaderException` and `FencedEpochException`. In the context of `scheduleAppend`, they mean the same thing to the caller. They both say that the raft state has transitioned and the operation is no longer possible. I'm inclined to keep `NotLeaderException` and document that it covers both cases. What do you think?

Reply:
Here I referred to the practices of `KafkaController` in which we have `NOT_CONTROLLER` and `STALE_CONTROLLER_EPOCH` for RPC errors, but here I also think it's better to consolidate `NotLeaderException` and `FencedEpochException` now.